Rewrite autograd producer consumer stream sync logic

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #151079

Also see previous work https://github.com/pytorch/pytorch/pull/142097

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k

soulitzer committed:
Rewrite autograd producer consumer logic

[ghstack-poisoned]

soulitzer committed:
Update on "Rewrite autograd producer consumer logic"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

soulitzer committed:
Update on "Rewrite autograd producer consumer logic"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

soulitzer commented on torch/csrc/autograd/input_buffer.cpp:
```diff
@@ -126,99 +126,134 @@ static void accumulate(
+  if (current_index == 0) {
+    if (is_accelerator) {
+      // Try assuming that these devices all have non-opt streams
+      TORCH_INTERNAL_ASSERT(opt_consumer_stream && opt_producer_stream)
```
The previous code didn't assume is_accelerator=True implies all stream are non-nullopt

soulitzer commented on torch/csrc/autograd/input_buffer.cpp:
```diff
@@ -126,99 +126,134 @@ static void accumulate(
+  if (num_dependencies == 1) {
+    if (is_accelerator) {
+      _wait_stream(opt_consumer_stream, opt_producer_stream, device_type);
+      record_stream_any_impl(var, *opt_consumer_stream);
```
May need extra tests to catch missing record streams

soulitzer committed:
Update on "Rewrite autograd producer consumer stream sync logic"


Also see previous work https://github.com/pytorch/pytorch/pull/142097

More details on design here WIP

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

soulitzer committed:
Update on "Rewrite autograd producer consumer stream sync logic"


Also see previous work https://github.com/pytorch/pytorch/pull/142097

More details on design here WIP

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

soulitzer commented on torch/csrc/autograd/input_buffer.cpp:
```diff
@@ -126,99 +126,138 @@ static void accumulate(
+          opt_accum_streams[pos] = opt_producer_stream;
+        } else {
+          c10::impl::VirtualGuardImpl guard(device_type);
+          opt_accum_streams[pos] = guard.getDefaultStream(*device);
```
Use the current stream instead?

guangyey commented on torch/csrc/autograd/input_buffer.cpp:
```diff
@@ -126,99 +126,138 @@ static void accumulate(
-        opt_accumulate_stream =
-            guard.getDefaultStream(opt_producer_stream->device());
-        opt_sync_stream = opt_producer_stream;
+  bool is_accelerator =
```
Hi, could you use [at::accelerator::isAccelerator](https://github.com/pytorch/pytorch/blob/d94cc0e9942087180305061dd693adff93448d2e/aten/src/ATen/DeviceAccelerator.cpp#L57) instead? It would help include other accelerators such as XPU and provide broader coverage.

soulitzer replied:
Yeah that sounds good to me, though I was planning to update this logic in the next PR just to keep each one as simple as possible

guangyey replied:
Sounds great! Thanks.

soulitzer committed:
Update on "Rewrite autograd producer consumer stream sync logic"


Also see previous work https://github.com/pytorch/pytorch/pull/142097

More details on design here WIP

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

soulitzer committed:
Update on "Rewrite autograd producer consumer stream sync logic"


Also see previous work https://github.com/pytorch/pytorch/pull/142097

More details on design here WIP

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

soulitzer committed:
Update on "Rewrite autograd producer consumer stream sync logic"


Also see previous work https://github.com/pytorch/pytorch/pull/142097

More details on design here WIP

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

soulitzer committed:
Update on "Rewrite autograd producer consumer stream sync logic"


Also see previous work https://github.com/pytorch/pytorch/pull/142097

More details on design here WIP

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

soulitzer committed:
Update on "Rewrite autograd producer consumer stream sync logic"


Also see previous work https://github.com/pytorch/pytorch/pull/142097

More details on design here WIP

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

soulitzer committed:
Update on "Rewrite autograd producer consumer stream sync logic"


Also see previous work https://github.com/pytorch/pytorch/pull/142097

More details on design here WIP

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

soulitzer commented on test/test_nestedtensor.py:
```diff
@@ -8377,16 +8377,6 @@ def f(values, offsets):
-            sample.kwargs.get("device", None) == "cpu"
-        ),
-        name="broken_to_backward",
-    ),
```
We no longer fail because we no longer call record_stream for case 3/ .to case. 
We want to call var.record_stream(accum_stream) when the accum stream is different from the "sync stream"
The "sync stream" prior to this PR was the default stream on the var device. With this PR, we consider the "sync stream" to be the producer stream.

albanD commented on torch/csrc/autograd/input_buffer.cpp:
```diff
@@ -129,96 +149,137 @@ static void accumulate(
-        const auto guard = c10::impl::VirtualGuardImpl(device_type);
-        record_stream_any_impl(var, *opt_accumulate_stream);
+        // Case 4/5
+        if (opt_producer_stream->device() == *device) {
```
Please make these 3 if statement flat

albanD commented on torch/csrc/autograd/input_buffer.cpp:
```diff
@@ -129,96 +149,137 @@ static void accumulate(
+          opt_accum_streams[pos] = opt_producer_stream;
+        } else {
+          opt_accum_streams[pos] =
+              at::accelerator::getCurrentStream(device->index());
```
So we're going with the current stream on that device.
Can you give some context on why you chose this resolution?

albanD reviewed (commented):
tbh the logic is pretty hard to follow as written, I'm curious if we could refactor it a bit to make it simpler?

soulitzer commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -1229,7 +1237,6 @@ auto Engine::compute_dependencies(
 
   // Queue contains all nodes that will start propagating gradients.
   // We no longer have to expand functions that don't require grad.
-  auto& dependencies = task.dependencies_;
```
no real reason/stylistic reasons - since we're mutating the field of task directly, might as well make that more clear

soulitzer commented:
> tbh the logic is pretty hard to follow as written, I'm curious if we could refactor it a bit to make it simpler?

Open to any suggestions on how to structure it better, but the way I'm thinking about it today is
1) First you can divide the logic in terms of which iteration you are on (or whether you need to iterate at all.
2) Then then you can further divide into the regimes of 2/3 and 4/5 where you are either accumulating on the consumer or producer.

So if I'm trying to reason about an example, I first determine if should be in the 2/3 or 4/5 regime (for a given buffer you are either always the 2/3 regime or 4/5 regime). Then, I can read the corresponding sections from current_index==0, or current_index > 0.

Are there any particular situations you find hard to reason about as written?

soulitzer commented on torch/csrc/autograd/input_buffer.cpp:
```diff
@@ -129,96 +149,137 @@ static void accumulate(
+  // [ First producer ]
+  // In the multiple producer case, we need to handle accumulation.
+  // When we see the very first producer, we do a couple things:
+  // - Determine the accumulation stream:
```
Let me update the comments to clarify these cases
(Yeah maybe I want to update the numbering, for now I grouped 2 and 3 because they are handled identically, but used to be 2 separate cases in the old categorization. )

soulitzer commented on torch/csrc/autograd/input_buffer.cpp:
```diff
@@ -129,96 +149,137 @@ static void accumulate(
+      // Case 4/5
+      mb_wait_stream(accum_stream, opt_producer_stream, device_type);
+      mb_record_stream_any_impl(
+          var, *accum_stream, /*prev_stream=*/*opt_producer_stream);
```
this can be deduped, we always want to have the accum wait for the new producer in both the 2/3 and 4/5 cases.

albanD commented on torch/csrc/autograd/input_buffer.cpp:
```diff
@@ -129,96 +149,137 @@ static void accumulate(
+  // In the multiple producer case, we need to handle accumulation.
+  // When we see the very first producer, we do a couple things:
+  // - Determine the accumulation stream:
+  //    case 1)   non-accelerator             -> n/a
```
This needs a legend lol

soulitzer replied:
Yeah I'm working on beefing up these comments haha

soulitzer commented:
Also let me try to make some simple simplifications/deduping first, I think I see some opportunities

soulitzer committed:
Update on "Rewrite autograd producer consumer stream sync logic"


Also see previous work https://github.com/pytorch/pytorch/pull/142097

More details on design here WIP

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

soulitzer committed:
Update on "Rewrite autograd producer consumer stream sync logic"


Also see previous work https://github.com/pytorch/pytorch/pull/142097

More details on design here WIP

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/input_buffer.cpp:
```diff
@@ -129,96 +149,147 @@ static void accumulate(
 
-  auto& old_var = buffer[pos];
-  if (!old_var.defined()) {
+  // Handle the case where var is on accelerator but producer node has no
```
```suggestion
  // Handle the case where var is on an accelerator but producer node has no
```

albanD commented on torch/csrc/autograd/input_buffer.cpp:
```diff
@@ -129,96 +149,147 @@ static void accumulate(
+  TORCH_INTERNAL_ASSERT(opt_consumer_stream && opt_producer_stream);
+
+  //
+  // [ Single producer case ]
```
You need to define what this means

albanD commented on torch/csrc/autograd/input_buffer.cpp:
```diff
@@ -129,96 +149,147 @@ static void accumulate(
+  //    case 1) var's device matches consumer node's canonical device
+  //            (The producer node's canonical device may or may not match)
+  //            -> accumulator stream = consumer stream
+  //    case 2) var's device only matches producer node's canonical device
```
Not sure what this "only" means here.
You mean it doesn't match the consumer but does match the producer?

soulitzer replied:
Yeah, will update comment to clarify

albanD commented on torch/csrc/autograd/input_buffer.cpp:
```diff
@@ -129,96 +149,147 @@ static void accumulate(
+  // First producer
+  // ==============
+  //
+  // A. Determine the accumulation stream:
```
What is A/B/C here?

soulitzer replied:
Probably not too useful. Removing.

albanD commented on torch/csrc/autograd/input_buffer.cpp:
```diff
@@ -129,96 +149,147 @@ static void accumulate(
+  // C. Accumulate var into buffer
+  // D. (Cases 2 and 3 only) If we are the last producer, we need to have
+  //    consumer wait for the accumulation stream. For case 1 this is
+  //    not necessary because the consumer stream is the accumulation stream.
```
You have special casing to skip same stream sync so this special case is not needed and it will just be a noop right?
Just to remove a conditional from the code below

albanD commented on torch/csrc/autograd/input_buffer.cpp:
```diff
@@ -129,96 +149,147 @@ static void accumulate(
+  // "test_side_stream_backward_overlap".
+  //
+  // There are two pieces of logic that happen:
+  //    Part 1: At the first iteration, stash the producer stream
```
You should record the event now and delay the wait later to maximize overlap possibilities right?

soulitzer replied:
Yeah, good catch

albanD commented on torch/csrc/autograd/input_buffer.cpp:
```diff
@@ -129,96 +149,147 @@ static void accumulate(
+    buffer[pos] = std::move(var);
+  } else {
+    // [ Nth producer ]
+    auto accum_stream = opt_accum_streams[pos];
```
nit: internal assert this was already set

albanD commented on torch/csrc/autograd/input_buffer.cpp:
```diff
@@ -129,96 +149,147 @@ static void accumulate(
-  //  with the consumer or producer.
-  //      Accumulation happens on the var device's default stream.
-
+  int current_index = accum_counts[pos]++;
```
`current_index` doesn't feel like a very descriptive name. It makes conditions below a bit hard to read.

soulitzer replied:
Renaming to `curr_producer_idx`

albanD reviewed (commented):
Much easier to read already !

soulitzer commented on torch/csrc/autograd/input_buffer.cpp:
```diff
@@ -129,96 +149,147 @@ static void accumulate(
+  //    the new producer stream
+  // B. Record stream (var will be used on the accum stream)
+  // C. Accumulate var into buffer
+  // D. (Cases 2 and 3 only) If we are the last producer, we need to have
```
Good point, but this is requires changes outside of InputBuffer.add

soulitzer commented on torch/csrc/autograd/input_buffer.h:
```diff
@@ -38,6 +44,15 @@ struct InputBuffer {
+  // Remember the first producer stream so we can delay synchronizing with
+  // it until the second producer is seen.
+  // See Note: [Delay synchronizing the first producer]
+  std::vector<std::optional<c10::Stream>> opt_first_producer_streams;
```
In the case where the consumer node's canonical device is the same as the gradient's device, the accumulation stream is the consumer stream rather than the producer stream.

soulitzer committed:
Update on "Rewrite autograd producer consumer stream sync logic"


Also see previous work https://github.com/pytorch/pytorch/pull/142097

More details on design here WIP

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

soulitzer committed:
Update

[ghstack-poisoned]

soulitzer commented on torch/csrc/autograd/input_buffer.cpp:
```diff
@@ -20,55 +20,6 @@
-            false, "Unknown layout in record_stream_any_impl");
-    }
-  }
-}
```
Code movement

soulitzer replied:
To make it possible to use in engine.cpp (related discussion https://github.com/pytorch/pytorch/pull/151079#discussion_r2079914220)

soulitzer committed:
Update

[ghstack-poisoned]

soulitzer commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -1063,13 +1063,33 @@ void Engine::evaluate_function(
+    if (opt_parent_stream.value() != inputs.ready_streams[pos].value()) {
+      // NOLINTNEXTLINE(bugprone-unchecked-optional-access)
+      opt_parent_stream->wait(inputs.ready_events[pos].value());
+      _record_stream_any_impl(inputs.buffer[pos], opt_parent_stream.value());
```
I might be missing:
```
if (opt_parent_stream->device() == device) {
  _record_stream_any_impl(inputs.buffer[pos], opt_parent_stream.value());
}
```

soulitzer replied:
Some general discussion on record stream correctness:

At a high level, for correctness, we want to "record stream" on every stream the it used on.

We consider the 5 cases:
1) Not accelerator
2) Device of grad matches consumer and producer device
3) Device of grad only matches consumer device
4) Device of grad only matches producer device
5) Device of grad matches neither

Case 1 is irrelevant to record stream. For the cases that do matter, record stream needs to happen in two places:

Record stream for consumer stream (engine.cpp)
- Case 4 & 5: 
  - No need, because we don't actually know which stream the user
    will end up using the gradient on. It would be the user's responsibility.
  
- Case 2 & 3:
  Record stream if the "ready stream" is different from the consumer stream. 
   - There is two cases:
        i) Either we did accumulate (due to multiple producers), the ready stream is the accumulation stream which is also the consumer stream.  (no need to record stream since we already allocated the tensor on the consumer stream)
        ii) Otherwise, the ready stream is the producer stream; which may or may  not be the consumer stream. (may or may not need to record stream)

Record stream for accumulation (input_buffer.cpp)
- Case 2 & 4
  - We assume the cuda caching allocator knows about the tensor on the producer stream, we may or may not record stream again depending on whether it matches. In theory, it doesn't hurt to record stream when it isn't needed. In practice, however, not all tensors support record stream due to dataptr access, so we try not to break them.
- Case 3 & 5
  - We don't know what stream the gradient was produced on so we should always record stream.

albanD commented on torch/csrc/autograd/input_buffer.cpp:
```diff
@@ -126,101 +78,195 @@ static void accumulate(
+    // 3)
+    ready_events[pos] = c10::Event{device_type};
+    ready_events[pos]->record(*opt_producer_stream);
+    ready_streams[pos] = opt_producer_stream;
```
I would argue you don't need to record this "ready_streams" and just always sync on "redy_events" when it's there.

@ngimel should we be concerned about the perf impact of making a stream waiting on itself?

soulitzer replied:
This would be a good simplification indeed, although even if making a stream on waiting on itself were free, you may need a conditional on "ready_streams" in order to avoid extra record_stream. The issue is not that there would be any perf impact I think (it should be a no-op), but rather because we need to access a tensors data ptr in order to record stream and that is not always possible, e.g. for wrapper subclasses (unless there is a better way?).

ngimel replied:
We'd rather avoid it for simple one-stream workloads which is still the majority

ngimel replied:
Oh and we should definitely avoid extra record_streams

albanD replied:
Ok!

albanD commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -1063,13 +1063,33 @@ void Engine::evaluate_function(
+    }
+    TORCH_INTERNAL_ASSERT(inputs.ready_events[pos].has_value());
+    TORCH_INTERNAL_ASSERT(inputs.ready_streams[pos].has_value());
+    TORCH_INTERNAL_ASSERT(opt_parent_stream.has_value());
```
That one is definitely not true! See your DtoH comment in input_buffer

soulitzer replied:
Good catch, whoops

soulitzer replied:
Wait no, parent stream is actually the consumer stream (I was confused when I replied and thought the parent stream was the producer stream). The /producer/ stream can be null, but the consumer is always non-null.

albanD commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -1063,13 +1063,33 @@ void Engine::evaluate_function(
+    TORCH_INTERNAL_ASSERT(inputs.ready_streams[pos].has_value());
+    TORCH_INTERNAL_ASSERT(opt_parent_stream.has_value());
+    // NOLINTNEXTLINE(bugprone-unchecked-optional-access)
+    if (opt_parent_stream.value() != inputs.ready_streams[pos].value()) {
```
Similar to the question to Natalia below. If we say we're happy with the noop sync, this will get heavily simplified.

soulitzer replied:
Resolving per https://github.com/pytorch/pytorch/pull/151079#discussion_r2080148805

albanD commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -1063,13 +1063,33 @@ void Engine::evaluate_function(
+    if (opt_parent_stream.value() != inputs.ready_streams[pos].value()) {
+      // NOLINTNEXTLINE(bugprone-unchecked-optional-access)
+      opt_parent_stream->wait(inputs.ready_events[pos].value());
+      _record_stream_any_impl(inputs.buffer[pos], opt_parent_stream.value());
```
I would assume all record stream to be done in input buffer and never needed here.

soulitzer replied:
It's possible to do it all in input buffer (PR updated), but it felt a little more intuitive to do here, e.g. I want to record_stream my var on a new stream everytime I use it on a new stream (which is before accumulation or before consumer node's execution). Curious what the reason is prefer to do it input_buffer?

soulitzer replied:
See https://github.com/pytorch/pytorch/pull/151079/commits/d0fa4cd0803dd6fc9d689ab36d4d32ff629ad605

soulitzer commented on torch/csrc/autograd/input_buffer.cpp:
```diff
@@ -126,101 +78,195 @@ static void accumulate(
+//            -> accumulator stream = producer stream
+//    case C) var device matches neither
+//            -> accumulator stream = var device's current stream
+//            See Note [Stream sync contract when dealing with
```
These are actually two different notes back to back!
Going to add a small gap between the two, hopefully it would make it clearer that they are two separate notes lol

soulitzer commented on torch/csrc/autograd/input_buffer.cpp:
```diff
@@ -126,101 +78,195 @@ static void accumulate(
+// 2) Accumulation on the accumulation straem
+// 3) Update the ready event and stream for the current position.
+//
+// NOLINTBEGIN(bugprone-unchecked-optional-access)
```
I tried haha, will experiment some more, but its very finnicky

albanD replied:
My experience is that it is very strict indeed. But for good reasons !
You definitely assume in a few place below that optionals are non-empty. But nothing really guarantees it. So internal asserts are good to have!

soulitzer replied:
So the issue was that linter gets confused when we are indexing a vector, e.g.  the linter doesn't learn from `TORCH_INTERNAL_ASSERT(ready_streams[pos])` that `ready_streams[pos]` is not null!
You need to assign `auto& ready_stream = ready_streams[pos]` and then `TORCH_INTERNAL_ASSERT(ready_streams)`

soulitzer commented on torch/csrc/autograd/input_buffer.cpp:
```diff
@@ -126,101 +146,173 @@ static void accumulate(
+  //
+  if (!is_accelerator) {
+    if (curr_producer_idx == 0) {
+      TORCH_INTERNAL_ASSERT(!buffer[pos].defined())
```
Nothing. This is an outdated version, I already changed it back!

soulitzer commented on torch/csrc/autograd/input_buffer.cpp:
```diff
@@ -126,101 +78,195 @@ static void accumulate(
+// The actual wait and record stream happens prior to the consumer's
+// execution. The logic here is mainly responsible for handling the
+// synchronization needed for accumulation and recording the event that the
+// consumer should wait on later.
```
Also see https://github.com/pytorch/pytorch/pull/151079#discussion_r2079816387 lol
Hopefully it would be clear enough that way.

soulitzer commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -1063,13 +1063,33 @@ void Engine::evaluate_function(
+      continue;
+    }
+    const auto device = inputs.buffer[pos].device();
+    // TODO: Use at::accelerator::isAccelerator(device->type()) instead
```
I guess I don't feel strongly about this being a TODO, but it does change behavior, e.g. enable more devices like MPS to use the accelerator path, so I wanted to split it out to keep things simple. Let me know if you'd prefer to combine it here.

albanD replied:
Ho right, I forgot it changes behavior. Sounds good!

soulitzer commented on torch/csrc/autograd/input_buffer.cpp:
```diff
@@ -126,101 +78,195 @@ static void accumulate(
+//
+// Note: [Autograd Producer-Consumer Stream Syncs]
+//
+// The actual wait and record stream happens prior to the consumer's
```
Added a intro sentence and refactored this section slightly!

soulitzer replied:
I guess I mention it because one might expect all the logic to be in InputBuffer haha

soulitzer committed:
Update

[ghstack-poisoned]

soulitzer committed:
Update

[ghstack-poisoned]

soulitzer committed:
Update

[ghstack-poisoned]

soulitzer commented on torch/csrc/autograd/input_buffer.cpp:
```diff
@@ -126,101 +78,195 @@ static void accumulate(
 
-  auto& old_var = buffer[pos];
-  if (!old_var.defined()) {
+  TORCH_INTERNAL_ASSERT(opt_consumer_stream && opt_producer_stream);
```
good guess, but it was actually https://github.com/pytorch/pytorch/pull/151079#discussion_r2080518132

albanD commented on torch/csrc/autograd/input_buffer.h:
```diff
@@ -38,6 +42,13 @@ struct InputBuffer {
+  // The events you need to wait for to ensure the corresponding buffers
+  // are ready. The events are updated as we accumulate into the buffer.
+  std::vector<std::optional<c10::Event>> ready_events;
+  // The streams corresponding to the events above.
```
nit: you can mention here that this is only used to check if more synchronization is needed or not.

albanD reviewed (approved):
Sounds great! Thanks!

soulitzer committed:
Update

[ghstack-poisoned]

soulitzer commented:
@pytorchbot merge

jeanschmidt commented:
@pytorchbot revert -m "Seems to have introduced regressions in internal signals, see [D74648937](https://www.internalfb.com/diff/D74648937)" -c ghfirst

soulitzer committed:
Update

[ghstack-poisoned]

soulitzer commented:
@pytorchbot merge

soulitzer commented:
@pytorchbot merge -i
